### PR TITLE
refactor: v8::Isolate::snapshot_creator instead of v8::SnapshotCreator

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit removes "v8::SnapshotCreator" in favor of
"v8::Isolate::snapshot_creator" API. All methods from
"v8::SnapshotCreator" are now available on the "v8::Isolate".
If a method is called and isolate was not set up as a snapshot
creator then it will panic.

This commit gets rid of long standing unsoundness in typing
that allowed to easily induce a panic or segfault if a user didn't
use snapshot creator API properly.

### CODE REVIEW

File: src/snapshot.rs [hunk 1]
```diff
@@ -164,10 +164,6 @@ impl SnapshotCreator {
     &mut self,
     function_code_handling: FunctionCodeHandling,
   ) -> Option<StartupData> {
-    {
```
Comment (bartlomieju) [inline src/snapshot.rs:167]: Some of the APIs in this file should be made private

File: src/isolate.rs [hunk 2]
```diff
@@ -1102,6 +1116,57 @@ impl Isolate {
     let arg = &mut callback as *mut F as *mut c_void;
     unsafe { v8__HeapProfiler__TakeHeapSnapshot(self, trampoline::<F>, arg) }
   }
+
+  /// Set the default context to be included in the snapshot blob.
+  /// The snapshot will not contain the global proxy, and we expect one or a
+  /// global object template to create one, to be provided upon deserialization.
+  ///
+  /// Panics
```
Comment (bartlomieju) [inline src/isolate.rs:1124]:
```suggestion
  /// # Panics
```

File: src/snapshot.rs [hunk 3]
```diff
@@ -89,26 +89,37 @@ pub enum FunctionCodeHandling {
...
-    }
+    };
+
+    let isolate_ptr =
+      unsafe { v8__SnapshotCreator__GetIsolate(&snapshot_creator) };
+    let mut owned_isolate = OwnedIsolate::new(isolate_ptr);
+    ScopeData::new_root(&mut owned_isolate);
+    owned_isolate.create_annex(Box::new(()));
+    owned_isolate.set_slot(snapshot_creator);
+    owned_isolate
```
Comment (bartlomieju) [inline src/snapshot.rs:122]: This is needed for https://github.com/denoland/rusty_v8/pull/973 where we will store allocations of existing blob

File: src/isolate.rs [hunk 4]
```diff
@@ -1303,6 +1398,29 @@ impl AsMut<Isolate> for Isolate {
...
+  ///
+  /// Panics if the isolate was not created using [`Isolate::snapshot_creator`]
+  #[inline(always)]
+  pub fn create_blob(
+    mut self,
+    function_code_handling: FunctionCodeHandling,
+  ) -> Option<StartupData> {
+    let mut snapshot_creator =
+      self.get_annex_mut().maybe_snapshot_creator.take().unwrap();
+    ScopeData::get_root_mut(&mut self);
```
Comment (piscisaureus) [inline src/isolate.rs:1415]: This call is unnecessary - the subsequent clear_scope_and_annex() call takes care
if it.

**Review by piscisaureus — APPROVED** [review]
Awesome! this was long overdue

File: src/isolate.rs (hunk 4 above)
Comment (bartlomieju) [reply src/isolate.rs:1415]: Did that in https://github.com/denoland/rusty_v8/pull/973

**Pull request merged**